ansible_base/rbac/models.py

```diff
@@ -0,0 +1,397 @@
+            existing_partials[permission_partial.obj_perm_id()] = permission_partial
+
+        expected_evaluations = self.expected_direct_permissions(types_prefetch)
+
+        for team in self.provides_teams.all():
+            for team_role in team.has_roles.all():
+                expected_evaluations.update(team_role.expected_direct_permissions(types_prefetch))
+
+        existing_set = set(existing_partials.keys())
+
+        to_delete = set()
+        for identifier in existing_set - expected_evaluations:
+            to_delete.add(existing_partials[identifier].id)
+
+        to_add = []
+        for codename, ct_id, obj_id in expected_evaluations - existing_set:
+            to_add.append(RoleEvaluation(codename=codename, content_type_id=ct_id, object_id=obj_id, role=self))
+
+        return (to_delete, to_add)
+
+
+# COMPUTED DATA
+class RoleEvaluation(models.Model):
+    """
+    Cached data that shows what permissions an ObjectRole gives its owners
+    example:
+        ObjectRole 423 gives users execute access to job template 37
+
+    RoleAncestorEntry model in old AWX RBAC system is a direct analog
+
+    This is used to make permission evaluations via querysets returning object ids
+    the data in this table is created from the ObjectRole and RoleDefinition data
+      you should not interact with this table yourself
+      the only method that should ever write to this table is
+        compute_object_role_permissions()
+
+    In the above example, "ObjectRole 423" may be a role that grants membership
+    to a team, and that team was given permission to another ObjectRole.
+    """
+
+    class Meta:
+        app_label = 'dab_rbac'
+        verbose_name_plural = _('role_object_permissions')
+        indexes = [
+            models.Index(fields=["role", "content_type_id", "object_id"]),  # used by get_roles_on_resource
+            models.Index(fields=["role", "content_type_id", "codename"]),  # used by accessible_objects
+        ]
+        constraints = [models.UniqueConstraint(name='one_entry_per_object_permission_and_role', fields=['object_id', 'content_type_id', 'codename', 'role'])]
+
+    def __str__(self):
+        return (
+            f'RoleEvaluation(pk={self.id}, codename={self.codename}, object_id={self.object_id}, '
+            f'content_type_id={self.content_type_id}, role_id={self.role_id})'
+        )
+
+    def save(self, *args, **kwargs):
+        if self.id:
+            raise RuntimeError('RoleEvaluation model is immutable and only used internally')
+        return super().save(*args, **kwargs)
+
+    role = models.ForeignKey(
+        ObjectRole, null=False, on_delete=models.CASCADE, related_name='permission_partials', help_text=_("The object role that grants this form of permission")
+    )
+    codename = models.TextField(null=False, help_text=_("The name of the permission, giving the action and the model, from the Django Permission model"))
```

Comment:
Why use `codename` and not a foreign key to the permission model here?

Reply:
The problem we're starting out with is that we need to construct a queryset, and the input we're getting is something like `JobTemplate.accessible_objects(user, 'view_jobtemplate')`, and I added aliasing in so that `JobTemplate.accessible_objects(user, 'view')` works, which is how I predict this is always going to work in practice.

This is little different from AWX today

https://github.com/ansible/awx/blob/f69f600cffa9757f961c222a38a82c58378f6a3a/awx/api/views/mixin.py#L151

which, in practice, does `JobTemplate.accessible_objects(self.request.user, 'read_role')`. Existing practices for that were copied over. So the action is taken in as a string and it's filtered that way. It would make sense to me to strip the model name actually, since that's redundant with `content_type_id`.

Reply:
Does it make more sense to use a foreign key now that we have a concrete permission model?

I'm worried that codenames could change if someone renames a model.

Reply:
Renaming a model is unsolved for `auth.Permission`, documented in https://github.com/ansible/django-ansible-base/issues/227

If you rename a model registered with RBAC, it will change the "evaluation" entries in its post_migrate signal. I'm not worried about that.